README.md
ccbingo
====

conference call bingo

to install:

```r
# install.packages('devtools')
devtools::install_github('raredd/ccbingo')
```

---

### play locally

**with built-in cards**

```r
library('ccbingo')
head(cards$zoom, 5)
```

```
## [1] "___, are you there?"            "*animal noises*"                                   "*child noises*"
## [4] "*loud, painful echo/feedback*"  "*sound of someone typing, possibly with a hammer*"
```

**start an interactive game**

```r
ccbingo(cards$zoom)
```

![](inst/shiny/img/ccbingo-local.png)

**customize with new cards, aesthetics, rules**

```r
ccbingo(
  rownames(mtcars), wins = function(x) if (sum(x) >= 6) '6 or more' else '',
  title = 'Monday lab meeting', free = '<3',
  col.grid = rainbow(6), col.free = 'white', col.select = 'white'
)
```

![](inst/shiny/img/ccbingo-custom.png)

```
## You win!
##
## --> 6 or more
##
## Fake an injury and end the call
```

### shiny app

```r
ccbingo_app()
```

![](inst/shiny/img/ccbingo-app.png)

---

### source

![](inst/shiny/img/ccbingo.png)


NEWS
ccbingo 0.0.1
--------
  * initial package
  * minimal shiny app


R/cards.R
### default cards
# cards
###


#' @format
#' \code{cards} is a \emph{named} list (by topic) of strings to use for cards.
#' 
#' @rdname ccbingo
#' @export
cards <- list(
  phone = c(
    'Am I on speaker?',
    'Are you on speaker?',
    'Can you take me off speaker?'
  ),
  
  skype = c(
    'Who uses skype?'
  ),
  
  webex = c(
    'My battery is dying',
    'webex crashed'
  ),
  
  zoom = c(
    '___, are you there?',
    '*animal noises*',
    '*child noises*',
    '*loud, painful echo/feedback*',
    '*sound of someone typing, possibly with a hammer*',
    'Can everyone go on mute?',
    'Can everyone see my screen?',
    'Can everyone stay on five more minutes?',
    'Can we take this offline?',
    'Can you email that to everyone?',
    'Can you hear me?',
    'Can you share your screen?',
    'Hello? Hello?',
    'Hey, guys, I have to jump to another call',
    'Hi, who just joined?',
    'I have a bad connection',
    'I have a hard stop at...',
    'I think there\'s a lag',
    'I think you\'re on mute',
    'I\'ll have to get back to you',
    'I\'m sorry, you cut out there',
    'It\'s still loading',
    'My connection is bad',
    'Next slide please',
    'So, *faded out* I can *unintelligible* by *cuts out*, ok?',
    'Sorry, didn\'t catch that, can you repeat?',
    'Sorry, go ahead',
    'Sorry, I was on mute',
    'Sorry, I\'m having connection issues',
    'Sorry, I\'m late... [lame excuse]',
    'Uh, ___, you\'re still sharing'
  )
)


inst/shiny/ui.R
library('shiny')

ui <- shinyUI(
  navbarPage(
    span(tagList(icon('video'), 'zoom bingo')),
    tabPanel(
      'play bingo',
      sidebarLayout(
        sidebarPanel(
          selectizeInput(
            'type', label = 'Select the card topic', selected = 'zoom',
            choices = names(cards)
          ),
          checkboxGroupInput(
            'wins', label = 'Ways to win:',
            c('Columns' = 'col', 'Rows' = 'row',
              'Diagonals' = 'diag', 'Corners' = 'corner'),
            selected = c('col', 'row', 'diag', 'corner')
          ),
          textInput('name', label = 'Who is playing?'),
          actionButton('submit', 'Play!')
        ),
        mainPanel(plotOutput('render_card', click = 'click'))
      )
    ),
    tabPanel('source', mainPanel(imageOutput('meme')))
  )
)


inst/shiny/globals.R
### shiny app functions
# cards, ccbingo_shiny, char_to_seed, check_for_win, bingo
###


cards <- list(
  phone = c(
    'Am I on speaker?',
    'Are you on speaker?',
    'Can you take me off speaker?'
  ),
  
  skype = c(
    'Who uses skype?'
  ),
  
  webex = c(
    'My battery is dying',
    'webex crashed'
  ),
  
  zoom = c(
    '___, are you there?',
    '*animal noises*',
    '*child noises*',
    '*loud, painful echo/feedback*',
    '*sound of someone typing, possibly with a hammer*',
    'Can everyone go on mute?',
    'Can everyone see my screen?',
    'Can everyone stay on five more minutes?',
    'Can we take this offline?',
    'Can you email that to everyone?',
    'Can you hear me?',
    'Can you share your screen?',
    'Hello? Hello?',
    'Hey, guys, I have to jump to another call',
    'Hi, who just joined?',
    'I have a bad connection',
    'I have a hard stop at...',
    'I think there\'s a lag',
    'I think you\'re on mute',
    'I\'ll have to get back to you',
    'I\'m sorry, you cut out there',
    'It\'s still loading',
    'My connection is bad',
    'Next slide please',
    'So, *faded out* I can *unintelligible* by *cuts out*, ok?',
    'Sorry, didn\'t catch that, can you repeat?',
    'Sorry, go ahead',
    'Sorry, I was on mute',
    'Sorry, I\'m having connection issues',
    'Sorry, I\'m late... [lame excuse]',
    'Uh, ___, you\'re still sharing'
  )
)

ccbingo_shiny <- function(card, seed, wins, clicks, ...) {
  seed <- if (is.character(seed))
    char_to_seed(seed) else as.integer(seed)
  
  if (is.null(clicks)) {
    set.seed(seed)
    bingo(card, NULL, ...)
    return(invisible(NULL))
  }
  
  if (inherits(clicks, 'list'))
    clicks <- do.call('cbind', clicks)
  sel <- t(apply(clicks, 1L, function(x) {
    x <- unname(findInterval(abs(x), 0:5) * sign(x) * c(1, -1))
    if (any(x > 5 | x < 1 | identical(x, c(3, 3))))
      c(NA, NA) else x
  }))
  sel <- na.omit(sel)
  
  dup <- apply(sel, 1L, toString)
  dup <- ave(seq_along(dup), dup, FUN = length)
  sel <- sel[dup %% 2 == 1, , drop = FALSE]
  
  set.seed(seed)
  mat <- bingo(card, sel, ...)
  win <- check_for_win(mat, wins)
  
  if (nzchar(win)) {
    set.seed(seed)
    bingo(card, sel, ...)
    text(grconvertX(0.5, 'ndc'), grconvertY(0.5, 'ndc'), 'w i n n e r !',
         col = 'red', family = 'serif', srt = 30, cex = 10, xpd = NA)
  }
  
  win
}

char_to_seed <- function(player) {
  x <- gsub('[^A-z0-9 ]', '', player)
  x <- strsplit(x, '')[[1L]]
  sum(as.integer(factor(x, c(' ', letters, LETTERS, 0:9))))
}

check_for_win <- function(m, wins = c('col', 'row', 'corner', 'diag')) {
  if (is.function(wins))
    return(wins(m))
  
  rn <- which(rowSums(m) == 5L)
  cn <- which(colSums(m) == 5L)
  lr <- all(diag(m))
  rl <- all(m[cbind(5:1, 1:5)])
  cr <- all(m[row(m) %in% c(1, 5) & col(m) %in% c(1, 5)])
  
  if (length(rn) && 'row' %in% wins)
    paste('row', rn)
  else if (length(cn) && 'col' %in% wins)
    paste('col', cn)
  else if (lr && 'diag' %in% wins)
    'major diagonal'
  else if (rl && 'diag' %in% wins)
    'minor diagonal'
  else if (cr && 'corner' %in% wins)
    'corners'
  else character(1L)
}

bingo <- function(card = '', select = NULL, title = NULL, free = NULL,
                  width = 25L, col.grid = c('grey95', 'white'),
                  col.select = 'lightyellow1', col.free = 'tomato') {
  `%||%` <- function(x, y) if (is.null(x)) y else x
  
  draw_card <- function(select = NULL, title, free, col.grid,
                        col.select, col.free) {
    plot(
      NA, type = 'n', xlim = c(0, 5), ylim = c(-5, 0),
      axes = FALSE, ann = FALSE
    )
    
    cc <- matrix(rep_len(col.grid, 25L), 5L, 5L)
    cc[select] <- col.select[1L]
    cc[3L, 3L] <- col.free[1L]
    
    co <- expand.grid(xleft = 0:4, ybottom = -(0:4), xright = 1:5, ytop = -(1:5))
    co <- c(co, list(col = cc))
    do.call('rect', co)
    
    op <- par(family = 'serif')
    text(2.5, -2.5, free %||% 'FREE', font = 2L, cex = 2)
    mt(title %||% 'CONFERENCE CALL', c(1, 5) - 0.5, cex = 1, line = 3)
    mt('BINGO', c(1, 5) - 0.5, cex = 3)
    par(family = op$family)
    
    gr <- matrix(FALSE, 5L, 5L)
    gr[3L, 3L] <- TRUE
    gr[!(cc %in% col.grid)] <- TRUE
    
    invisible(t(gr))
  }
  
  mt <- function(x, to, from = range(x), space = '', ...) {
    l <- cbind(strsplit(x, '')[[1L]], ' ')[, nzchar(space) + 1L]
    x <- seq_along(l)
    x <- (x - from[1L]) / diff(from) * diff(to) + to[1L]
    mtext(l, at = x, ...)
  }
  
  card <- sample(card, 25L, replace = length(card) < 25L)
  
  op <- par(mar = c(0, 0, 5, 0), family = 'serif')
  on.exit(par(op))
  
  gr <- draw_card(select, title, free, col.grid, col.select, col.free)
  sq <- sapply(card, function(x)
    paste0(strwrap(x, width), collapse = '\n'))
  
  co <- expand.grid(x = 1:5 - 0.5, y = -(1:5) + 0.5)
  co <- c(co, list(col = 1L, labels = replace(sq, 13L, '')))
  do.call('text', co)
  
  invisible(gr)
}


R/ccbingo.R
### ccbingo functions
# ccbingo, char_to_seed, ccbingo_app
#
# unexported
# ccbingo_shiny, check_for_win, bingo
###


#' Conference call bingo
#' 
#' @description
#' Play conference call bingo locally or with \code{\link[shiny]{shiny}}.
#' 
#' \code{ccbingo} runs in a regular \code{R} session and prompts the user for
#' each click using \code{\link{locator}}. \code{ccbingo:::ccbingo_shiny} has
#' similar functionality but requires a list of click coordinates.
#' 
#' \code{char_to_seed} creates a simple string-to-seed mapping which allows
#' cards to be reproduced from a player name, for example.
#' 
#' \code{cards} is a list of built-in phrases to use for cards.
#' 
#' @param card a vector of character strings to make random cards; strings
#' are randomly selected from \code{card} without replacement but recycled
#' as needed to draw a 5x5 grid
#' @param seed an integer passed to \code{\link{set.seed}} to reproduce cards;
#' alternatively, a character string to be converted to a seed with
#' \code{char_to_seed}
#' @param wins ways to win the game, any combination of \code{"col"},
#' \code{"row"}, \code{"corner"}, or \code{"diag"}; alternatively, a function
#' to determine wins returning \code{''} for no win and a non-zero length
#' string for wins
#' @param ... additional arguments passed to \code{ccbingo:::bingo} such as
#' \code{title} (card title), \code{free} (free space text), \code{width}
#' (text width per box), \code{col.grid} (grid colors), \code{col.select}
#' (color of selected boxes), or \code{col.free} (color of free box)
#' @param player player name
#' 
#' @examples
#' char_to_seed('joe')
#' char_to_seed('sue')
#' 
#' \dontrun{
#' ## these are not exported
#' ccbingo:::bingo()
#' ccbingo:::bingo(letters)
#' ccbingo:::bingo(
#'   cards$zoom, col.grid = rainbow(6), col.free = 'white',
#'   free = '<3', title = 'come and play'
#' )
#' 
#' ## play interactively in an r session
#' ccbingo(cards$zoom)
#' ccbingo(cards$zoom, title = 'my bingo', free = '<3', col.free = 'white',
#'         col.select = 'transparent', col.grid = rainbow(6))
#' 
#' 
#' ## same as above except pass a list of recorded clicks
#' 
#' ## interactive -- first set up a game then make clicks
#' ccbingo(cards$zoom)
#' clicks <- locator(5)
#' ccbingo:::ccbingo_shiny(cards$zoom, 1, NULL, clicks)
#' 
#' ## or pass a list of clicks directly to game
#' set.seed(1)
#' clicks <- list(x = runif(15, 0, 5), y = -runif(15, 0, 5))
#' ccbingo:::ccbingo_shiny(cards$zoom, 1, NULL, clicks)
#' 
#' ## run shiny app
#' ccbingo_app()
#' }
#' 
#' @name ccbingo

#' @rdname ccbingo
#' @export
ccbingo <- function(card, seed = 1L, wins = c('col', 'row', 'corner', 'diag'),
                    ...) {
  seed <- if (is.character(seed))
    char_to_seed(seed) else as.integer(seed)
  set.seed(seed)
  mat <- bingo(card, NULL, ...)
  sel <- NULL
  
  select_box <- function(coords = NULL) {
    co <- if (is.null(coords))
      unlist(locator(1L)) else coords
    if (is.null(co))
      return(NULL)
    
    co <- unname(findInterval(abs(co), 0:5) * sign(co) * c(1, -1))
    
    if (any(co > 5 | co < 1)) {
      message('Outside of grid -- select a box')
      co <- Recall()
    } else if (identical(co, c(3, 3))) {
      message('FREE space -- select a box')
      co <- Recall()
    }
    
    co
  }
  
  while (TRUE) {
    cur <- select_box()
    if (is.null(cur))
      return('game stopped')
    sel <- rbind(sel, cur)
    dup <- duplicated(sel) | duplicated(sel, fromLast = TRUE)
    sel <- sel[!dup, , drop = FALSE]
    
    set.seed(seed)
    mat <- bingo(card, sel, ...)
    win <- check_for_win(mat, wins)
    
    if (nzchar(win))
      break
  }
  
  cat(sprintf('You win!\n\n--> %s\n\nFake an injury and end the call\n', win))
  invisible(win)
}

#' @rdname ccbingo
#' @export
char_to_seed <- function(player) {
  x <- gsub('[^A-z0-9 ]', '', player)
  x <- strsplit(x, '')[[1L]]
  sum(as.integer(factor(x, c(' ', letters, LETTERS, 0:9))))
}

#' @rdname ccbingo
#' @export
ccbingo_app <- function() {
  path <- system.file('shiny', package = 'ccbingo')
  shiny::runApp(path, launch.browser = TRUE)
}

#' ccbingo for shiny
#' 
#' Similar to \code{\link{ccbingo}} but uses a list of click coordinates since
#' \code{\link[shiny]{shiny}} does not call \code{\link{locator}}.
#' 
#' @param card,seed,wins,... see \code{\link{ccbingo}}
#' @param clicks for the \code{\link[shiny]{shiny}} app, a list of click
#' locations (in \link[=grconvertX]{user coordinates})
#' 
#' @examples
#' \dontrun{
#' ## first set up a game then make clicks
#' ccbingo(cards$zoom)
#' clicks <- locator(5)
#' ccbingo:::ccbingo_shiny(cards$zoom, 1, NULL, clicks)
#' 
#' ## or pass a list of clicks directly to game
#' set.seed(1)
#' clicks <- list(x = runif(15, 0, 5), y = -runif(15, 0, 5))
#' ccbingo:::ccbingo_shiny(cards$zoom, 1, NULL, clicks)
#' }
#' 
#' @noRd

ccbingo_shiny <- function(card, seed, wins, clicks, ...) {
  seed <- if (is.character(seed))
    char_to_seed(seed) else as.integer(seed)
  
  if (is.null(clicks)) {
    set.seed(seed)
    bingo(card, NULL, ...)
    return(invisible(NULL))
  }
  
  if (inherits(clicks, 'list'))
    clicks <- do.call('cbind', clicks)
  sel <- t(apply(clicks, 1L, function(x) {
    x <- unname(findInterval(abs(x), 0:5) * sign(x) * c(1, -1))
    if (any(x > 5 | x < 1 | identical(x, c(3, 3))))
      c(NA, NA) else x
  }))
  sel <- na.omit(sel)
  
  dup <- apply(sel, 1L, toString)
  dup <- ave(seq_along(dup), dup, FUN = length)
  sel <- sel[dup %% 2 == 1, , drop = FALSE]
  
  set.seed(seed)
  mat <- bingo(card, sel, ...)
  win <- check_for_win(mat, wins)
  
  if (nzchar(win)) {
    set.seed(seed)
    bingo(card, sel, ...)
    text(grconvertX(0.5, 'ndc'), grconvertY(0.5, 'ndc'), 'w i n n e r !',
         col = 'red', family = 'serif', srt = 30, cex = 10, xpd = NA)
  }
  
  win
}

#' Checking for wins
#' 
#' @param m a 5x5 logical matrix of selected boxes
#' @param wins ways to win the game, any combination of \code{"col"},
#' \code{"row"}, \code{"corner"}, or \code{"diag"}; alternatively, a function
#' to determine wins returning \code{''} for no win and a non-zero length
#' string for wins
#' 
#' @examples
#' m <- matrix(FALSE, 5, 5)
#' diag(m) <- TRUE
#' m
#' ccbingo:::check_for_win(m)
#' ## dont count diagonal wins
#' ccbingo:::check_for_win(m, c('col', 'row', 'corner'))
#' 
#' six_boxes <- function(x) if (sum(x) > 5) 'winner' else ''
#' ccbingo:::check_for_win(m, six_boxes)
#' m[, 1] <- TRUE
#' ccbingo:::check_for_win(m, six_boxes)
#' 
#' @noRd

check_for_win <- function(m, wins = c('col', 'row', 'corner', 'diag')) {
  if (is.function(wins))
    return(wins(m))
  
  if (is.null(wins))
    wins <- c('col', 'row', 'corner', 'diag')
  
  rn <- which(rowSums(m) == 5L)
  cn <- which(colSums(m) == 5L)
  lr <- all(diag(m))
  rl <- all(m[cbind(5:1, 1:5)])
  cr <- all(m[row(m) %in% c(1, 5) & col(m) %in% c(1, 5)])
  
  if (length(rn) && 'row' %in% wins)
    paste('row', rn)
  else if (length(cn) && 'col' %in% wins)
    paste('col', cn)
  else if (lr && 'diag' %in% wins)
    'major diagonal'
  else if (rl && 'diag' %in% wins)
    'minor diagonal'
  else if (cr && 'corner' %in% wins)
    'corners'
  else character(1L)
}

#' Draw a bingo card
#' 
#' @param card a vector of character strings to make random cards; strings
#' are randomly selected from \code{card} without replacement but recycled
#' as needed to draw a 5x5 grid
#' @param select an optional matrix of selected locations
#' @param title,free strings for the main title and free space
#' @param width a positive integer giving the target column for wrapping
#' lines; see \code{\link{strwrap}}
#' @param col.grid,col.select,col.free colors for the grid, selected boxes,
#' and free space, respectively
#' 
#' @examples
#' bingo()
#' bingo(letters)
#' bingo(cards$zoom)
#' 
#' @noRd

bingo <- function(card = '', select = NULL, title = NULL, free = NULL,
                  width = 25L, col.grid = c('grey95', 'white'),
                  col.select = 'lightyellow1', col.free = 'tomato') {
  `%||%` <- function(x, y) if (is.null(x)) y else x
  
  draw_card <- function(select = NULL, title, free, col.grid,
                        col.select, col.free) {
    plot(
      NA, type = 'n', xlim = c(0, 5), ylim = c(-5, 0),
      axes = FALSE, ann = FALSE
    )
    
    cc <- matrix(rep_len(col.grid, 25L), 5L, 5L)
    cc[select] <- col.select[1L]
    cc[3L, 3L] <- col.free[1L]
    
    co <- expand.grid(xleft = 0:4, ybottom = -(0:4), xright = 1:5, ytop = -(1:5))
    co <- c(co, list(col = cc))
    do.call('rect', co)
    
    op <- par(family = 'serif')
    text(2.5, -2.5, free %||% 'FREE', font = 2L, cex = 2)
    mt(title %||% 'CONFERENCE CALL', c(1, 5) - 0.5, cex = 1, line = 3)
    mt('BINGO', c(1, 5) - 0.5, cex = 3)
    par(family = op$family)
    
    gr <- matrix(FALSE, 5L, 5L)
    gr[3L, 3L] <- TRUE
    gr[!(cc %in% col.grid)] <- TRUE
    
    invisible(t(gr))
  }
  
  mt <- function(x, to, from = range(x), space = '', ...) {
    l <- cbind(strsplit(x, '')[[1L]], ' ')[, nzchar(space) + 1L]
    x <- seq_along(l)
    x <- (x - from[1L]) / diff(from) * diff(to) + to[1L]
    mtext(l, at = x, ...)
  }
  
  card <- sample(card, 25L, replace = length(card) < 25L)
  
  op <- par(mar = c(0, 0, 5, 0), family = 'serif')
  on.exit(par(op))
  
  gr <- draw_card(select, title, free, col.grid, col.select, col.free)
  sq <- sapply(card, function(x)
    paste0(strwrap(x, width), collapse = '\n'))
  
  co <- expand.grid(x = 1:5 - 0.5, y = -(1:5) + 0.5)
  co <- c(co, list(col = 1L, labels = replace(sq, 13L, '')))
  do.call('text', co)
  
  invisible(gr)
}
